adam/perception/perception_graph_predicates.py

```diff
@@ -0,0 +1,644 @@
+    (if the are *attrs* classes, set *eq=False*).
+    """
+
+    @abstractmethod
+    def __call__(self, graph_node: PerceptionGraphNode) -> bool:
+        """
+        Determines whether a *graph_node* is matched by this predicate.
+        """
+
+    @abstractmethod
+    def dot_label(self) -> str:
+        """
+        Node label to use when rendering patterns as graphs using *dot*.
+        """
+
+    @abstractmethod
+    def is_equivalent(self, other: "NodePredicate") -> bool:
+        """
+        Compares two predicates and return true if they are equivalent
+        """
+
+    def matches_predicate(self, predicate_node: "NodePredicate") -> bool:
+        """
+        Determines whether a NodePredicate matches another Node Predicate
+        """
+
+
+@attrs(frozen=True, slots=True, eq=False)
+class AnyGraphNodePredicate(NodePredicate):
+    """
+    Matches any node of type `GraphNode`.
+    """
+
+    def __call__(self, graph_node: PerceptionGraphNode) -> bool:
+        return isinstance(graph_node, GraphNode)
+
+    def dot_label(self) -> str:
+        return "GraphNode(*)"
+
+    def is_equivalent(self, other) -> bool:
+        return isinstance(other, AnyGraphNodePredicate)
+
+    def matches_predicate(self, predicate_node: NodePredicate) -> bool:
+        return isinstance(predicate_node, AnyGraphNodePredicate)
+
+
+@attrs(frozen=True, slots=True, eq=False)
+class AnyObjectPredicate(NodePredicate):
+    """
+    Matches any node of type `ObjectNode`.
+    """
+
+    def __call__(self, graph_node: PerceptionGraphNode) -> bool:
+        return isinstance(graph_node, ObjectNode)
+
+    def dot_label(self) -> str:
+        return "ObjectNode(*)"
+
+    def is_equivalent(self, other) -> bool:
+        return isinstance(other, AnyObjectPredicate)
+
+    def matches_predicate(self, predicate_node: NodePredicate) -> bool:
+        return isinstance(predicate_node, AnyObjectPredicate)
+
+
+@attrs(frozen=True, slots=True, eq=False)
+class CategoricalPredicate(NodePredicate):
+    """
+    Matches any node of type `ObjectNode`.
+    """
+
+    value: str = attrib(validator=instance_of(str))
+
+    def __call__(self, graph_node: PerceptionGraphNode) -> bool:
+        if isinstance(graph_node, CategoricalNode):
+            return self.value == graph_node.value
+        return False
+
+    @staticmethod
+    def from_node(categorical_node: CategoricalNode) -> "CategoricalPredicate":
+        return CategoricalPredicate(value=categorical_node.value)
+
+    def dot_label(self) -> str:
+        return f"CategoryFeature(value={self.value})"
+
+    def is_equivalent(self, other) -> bool:
+        return isinstance(other, CategoricalPredicate)
+
+    def matches_predicate(self, predicate_node: NodePredicate) -> bool:
+        if isinstance(predicate_node, CategoricalPredicate):
+            return predicate_node.value == self.value
+        return False
+
+
+@attrs(frozen=True, slots=True, eq=False)
+class ContinuousPredicate(NodePredicate):
```

Comment:
Not sure what the distinction between Categorical and Continuous Node Predicates is. Does this have to do with continuous representations/weights?

Reply:
Yes that's correct. For any feature which is a continuous value we would use the continuous node but for categorical values (e.g. material) we'd use the category node. I'm unsure we currently plan to have a continuous feature but it was easy enough to add now just so its available in the future.

Reply:
I see. Hmm, it seems that all of our features (and corresponding predicates) would be continuous though, no? I thought even something "categorical" would actually have a weight associated with it (like in the weight field from the comment above)

Reply:
We've discussed binning color by just arbitrarily picking the breakpoints in RGB space rather than trying to explore the problem of "defining what 'red' is in RGB" and dealing with different lighting conditions, etc. So in that case color would become a categorical definition rather than continuous correct? For an example that's not color I suspect the feature of "material" will come to use as a set of discrete categories which we may end up referring to as 0, 1, 2, etc rather than by a human name but the pertained system probably only has a finite list of materials it can categorize to.

Yes I agree these nodes will also have a weight, I was under the impression this "weight" could be the confidence of the perception system that this feature is accurately detected or alternatively in the graph patterns the weight of importance for "if we are missing these feature can we still say this object matches this pattern" E.g. if an apple is matte rather than shiny is it still an Apple ? but we don't loose the information that maybe this was an outlier and actually most apples are shiny.

Reply:
Don't we have a specific predicate for colors though? Either way, these will be specified more as we go. I was just confused about what `ContinuousNodePredicate` was supposed to capture